---
title: "Endothelial GD expression"
author: "Ruiyu Ray Wang"
date: "`r Sys.Date()`"
output: html_document
knit: (function(input_file, encoding){
  out_dir <- '../docs';
  rmarkdown::render(input_file,
    encoding=encoding,
    output_file=file.path(dirname(input_file), out_dir, 'vanlandewijck_duan_GD_expr.html'))})
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
rm(list = ls(all.names = TRUE))

## Register font for plotting
# extrafont::font_import()
# extrafont::loadfonts(device="postscript")  ## Register fonts for Windows bitmap output
# extrafont::fonts()  ## vector of font family names

READ_CACHE = TRUE
```
```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(ggplot2)
  library(forcats)
  library(ggrepel)
  library(ggpubr)
  library(ggsignif)
  # library(gginnards)
  library(RColorBrewer)
  library(clusterProfiler)
  library(org.Mm.eg.db)
  library(scWidgets)
})
```

#### Vanlanderwijck

```{r}
vanlandewijck.seurat <- readRDS(file = "../data/Vanlandewijck_2018/Vanlandewijck.seurat.brain.rds")
```

```{r}
DimPlot(vanlandewijck.seurat, reduction = "tsne", label = TRUE) +
  NoAxes() +
  NoLegend()
```

```{r}
# Customize label
emb <- as.data.frame(Embeddings(vanlandewijck.seurat, reduction = "tsne"))
emb$cell_type <- vanlandewijck.seurat$cell_type
emb_avg <- emb %>% dplyr::group_by(cell_type) %>% dplyr::summarise(tSNE_1_avg = mean(tSNE_1),tSNE_2_avg = mean(tSNE_2),
                                                                   tSNE_1_med = median(tSNE_1),tSNE_2_med = median(tSNE_2))
```
```{r}
d_van <- DimPlot(vanlandewijck.seurat, reduction = "tsne", cols = brewer.pal(n = 8, name = "Set2")[c(2,1,3)]) +
  geom_label(data = emb_avg, aes(x = tSNE_1_med, y = tSNE_2_med), label = emb_avg$cell_type, size = 6, fill = "orange", 
             alpha = .3, fontface = "bold") +
  NoAxes() +
  NoLegend()
d_van
ggsave(filename = "t-SNE_Vanlandewijck.tiff", plot = d_van, device = "tiff", path = "../data/Vanlandewijck_2018/", dpi = "print", width = 5, height = 5)
```
```{r}
DimPlot(vanlandewijck.seurat, reduction = "tsne", cols = brewer.pal(n = 8, name = "Set2")[c(2,1,3)]) +
  NoAxes() +
  NoLegend()
ggsave(filename = "t-SNE_Vanlandewijck.clean.tiff", device = "tiff", path = "../data/Vanlandewijck_2018/", dpi = "print", width = 5, height = 5)
```

```{r}
reduction = 'tsne'
brewer.use = 'OrRd'

out_dir <- file.path('../data/Vanlandewijck 2018/',
                     paste('goi_featureplots', reduction, brewer.use,sep = "_"))
if(!dir.exists(out_dir)){dir.create(out_dir)}
genes_of_interest <- c("Casp1","Casp4","Gsdmd","Tlr4","Tlr3","Cd14","Lbp","Nlrp3","Pycard","Tnf","Il6","Ninj1","Irf2")

iter_feature_plot_save <- function(object, features_plot, reduction = NULL, brewer.use = NULL, out_dir){
  if(is.null(reduction)){reduction = "tsne"}
  if(is.null(brewer.use)){brewer.use = "OrRd"}
  
  for (feature in features_plot){
    # Original FeaturePlot
    p <- FeaturePlot(object, features = feature, reduction = reduction) +
      scale_colour_gradientn(colours = brewer.pal(n = 5, name = brewer.use))
    # FeaturePlot without legend, title, axes.
    p_clean <- FeaturePlot(object, features = feature, reduction = reduction) +
      ggtitle(label = NULL) + 
      NoAxes() + 
      theme(panel.border = element_rect(colour = "black", fill=NA, size=.5),
            legend.position = c(0.06, 0.24),
            legend.key.size = unit(0.55,'cm'),
            legend.text = element_blank()) + 
      scale_colour_gradientn(colours = brewer.pal(n = 5, name = brewer.use)) +
      # annotate(geom = "text", x = -38, y = 32, label = feature, fontface = 2, size = 6)
    
    ggsave(filename = paste0("featureplot_",feature,"_",reduction,".tiff"), plot = p, device = "tiff", path = out_dir, width = 4, height = 4)
    ggsave(filename = paste0("featureplot_",feature,"_",reduction,".clean.tiff"), plot = p_clean, device = "tiff", path = out_dir,
           width = 4, height = 4)
  }
}

iter_feature_plot_save(vanlandewijck.seurat, features_plot = genes_of_interest, reduction = reduction, out_dir = out_dir)
```

#### Duan

```{r}
duan.seurat <- readRDS(file = "../data/Duan_2018/Duan.seurat.annotated.rds")
duan.endo <- readRDS(file = "../data/Duan_2018/Duan.endo.rds")
```

```{r}
## t_SNE grouped by cell_type_brief
# duan.seurat <- reorderIdents(duan.seurat, 
#                              ident_type = "cell_type_brief", 
#                              desired_order = c("SMC","FB","Peri","RBCs","OPCs","Astro","Neurons","Endo",
#                                                "OL","CPEpiC","CR","MG","Interneurons","Epen","MΦ"))
d <- DimPlot(duan.seurat, group.by = "cell_type_brief", reduction = "tsne.int") + 
  ggtitle(label = NULL) +
  NoLegend() +
  NoAxes()
ggsave(filename = "t_SNE_int_ctb.tiff", plot = d, device = "tiff", path = "LPS Treatment/Duan 2018/figs", dpi = "print", width = 6, height = 5)

## t-SNE guide
cols <- brewer.pal(n = 12, name = "Paired")
d_guide_1 <- DimPlot(duan.seurat, group.by = "cell_type_brief", reduction = "tsne.int", label = TRUE, label.box = TRUE) +
  ggtitle(label = "Cell Type")
d_guide_2 <- DimPlot(duan.seurat, group.by = "treat", reduction = "tsne.int", cols = cols[c(10,7)]) +
  ggtitle(label = "Treatment")
d_guide <- d_guide_1 | d_guide_2
ggsave(filename = "t_SNE_guide.jpeg", plot = d_guide, device = "jpeg", path = "LPS Treatment/Duan 2018/figs", width = 16, height = 7)
```
```{r}
f_gd <- FeaturePlot(duan.seurat, features = "Gsdmd", reduction = "tsne.int") + NoAxes()
```

Test some variables.
```{r}
for(i in seq(400,800, by = 10)){
  duan.endo <- RunUMAP(duan.endo, dims = 1:30, n.epochs = i, umap.method = "umap-learn")
  dplot <- DimPlot(duan.endo, group.by = "treat") + NoLegend() + NoAxes()
  ggsave(filename = paste0("dplot_",i,"_epochs.jpeg"), plot = dplot, device = "jpeg", path = "LPS Treatment/Duan 2018/figs/test_epoch/", width = 4, height = 4)
}

for(i in seq(1,30)){
  duan.endo <- RunTSNE(duan.endo, dims = 1:30, seed.use = i)
  dplot <- DimPlot(duan.endo, reduction = "tsne", group.by = "treat") + NoLegend() + NoAxes()
  ggsave(filename = paste0("dplot_seed_",i,".jpeg"), plot = dplot, device = "jpeg", path = "LPS Treatment/Duan 2018/figs/test_epoch/", width = 4, height = 4)
}  # t-SNE with seed = 20 seems to give best look figure

duan.endo <- RunTSNE(duan.endo, dims = 1:30, seed.use = 20)
d_plot_use <- DimPlot(duan.endo, group.by = "treat", reduction = "tsne", cols = cols[c(10,7)]) + 
  annotate("rect", xmin = 18, xmax = 34, ymin = -17, ymax = -27, alpha = .05, fill = "orange") +
  ggtitle(label = NULL) +
  NoLegend() +
  NoAxes() +
  annotate("text", x = 25, y = -20, label = "LPS", color = cols[10], fontface = 2, size = 5) +
  annotate("text", x = 26, y = -24, label = "Saline", color = cols[7], fontface = 2, size = 5)
  # geom_rect(xmin = 20, xmax = 32, ymin = -17, ymax = -27, alpha = .5, color = "white", fill = "grey")
out_dir <- '~/Documents/Work/scRNAseq/EC & Single-Cell Sequencing/LPS Treatment/Duan 2018/figs/'
ggsave(filename = "dplot_endo.tiff", device = "tiff", path = out_dir, width = 4, height = 4, dpi = 300, plot = d_plot_use)

DimPlot(duan.endo, group.by = "treat", reduction = "tsne", cols = cols[c(10,7)]) + 
  ggtitle(label = NULL) +
  NoLegend() +
  NoAxes()
ggsave(filename = "dplot_endo.clean.tiff", device = "tiff", path = out_dir, width = 4, height = 4, dpi = 300)
```

```{r}
reduction = 'tsne'
brewer.use = 'RdPu'

out_dir <- file.path('LPS Treatment/Duan 2018/figs',
                     paste('goi_featureplots', reduction, brewer.use,sep = "_"))
if(!dir.exists(out_dir)){dir.create(out_dir)}
genes_of_interest <- c("Casp1","Casp4","Gsdmd","Tlr4","Tlr3","Cd14","Lbp","Nlrp3","Pycard","Il1b","Tnf","Il6","Ninj1","Irf2","Nlrc4","Aim2","Mefv")  ## "Nlrp1b",

iter_feature_plot_save <- function(object, features_plot, reduction = NULL, brewer.use = NULL, out_dir){
  if(is.null(reduction)){reduction = "tsne"}
  if(is.null(brewer.use)){brewer.use = "OrRd"}
  
  for (feature in features_plot){
    # Original FeaturePlot
    p <- FeaturePlot(object, features = feature, reduction = reduction) +
      scale_colour_gradientn(colours = brewer.pal(n = 5, name = brewer.use))
    # FeaturePlot without legend, title, axes.
    p_clean <- FeaturePlot(object, features = feature, reduction = reduction) +
      ggtitle(label = NULL) + 
      NoAxes() + 
      theme(panel.border = element_rect(colour = "black", fill=NA, size=.5),
            legend.position = c(0.82, 0.24),
            legend.key.size = unit(0.55,'cm'),
            legend.text = element_blank()) + 
      scale_colour_gradientn(colours = brewer.pal(n = 5, name = brewer.use))
      # annotate(geom = "text", x = -23, y = 25, label = feature, fontface = 2, size = 6)
    
    ggsave(filename = paste0("featureplot_",feature,"_",reduction,".tiff"), plot = p, device = "tiff", path = out_dir, width = 4, height = 4)
    ggsave(filename = paste0("featureplot_",feature,"_",reduction,".clean.tiff"), plot = p_clean, device = "tiff", path = out_dir,
           width = 4, height = 4)
  }
}

iter_feature_plot_save(duan.endo, features_plot = genes_of_interest, reduction = reduction, brewer.use = brewer.use, out_dir = out_dir)
```

Plotting Seurat images using `tiff` device in `ggsave()` started to crash Rstudio. (2021.11.17)
Reported to github: <https://github.com/satijalab/seurat/issues/5304>

For Nlrp1b, use a workaround.
```{r}
reduction = 'tsne'
brewer.use = 'RdPu'

out_dir <- file.path('~/DATA/WeiChao/EC & Single-Cell Sequencing/LPS Treatment/Duan 2018/figs',
                     paste('goi_featureplots', reduction, brewer.use,sep = "_"))
if(!dir.exists(out_dir)){dir.create(out_dir)}
feature = "Nlrp1b"

sub.endo <- subset(duan.seurat, cells = Cells(duan.endo))
sub.endo[["tsne"]] <- CreateDimReducObject(embeddings = Embeddings(duan.endo, reduction = "tsne"), assay = "RNA")
"Nlrp1b" %in% rownames(sub.endo)  # Now has the gene

# Original FeaturePlot
p <- FeaturePlot(sub.endo, features = feature, reduction = reduction, cols = c("#FBECE3","#FBECE3"))
# FeaturePlot without legend, title, axes.
p_clean <- FeaturePlot(sub.endo, features = feature, reduction = reduction, cols = c("#FBECE3","#FBECE3")) +
      ggtitle(label = NULL) + 
      NoAxes() + 
      theme(panel.border = element_rect(colour = "black", fill=NA, size=.5),
            legend.position = c(0.82, 0.24),
            legend.key.size = unit(0.55,'cm'),
            legend.text = element_blank())

ggsave(filename = paste0("featureplot_",feature,"_",reduction,".tiff"), plot = p, device = "tiff", path = out_dir, width = 4, height = 4)
ggsave(filename = paste0("featureplot_",feature,"_",reduction,".clean.tiff"), plot = p_clean, device = "tiff", path = out_dir, width = 4, height = 4)
```

#### Fig Violin Plot

```{r}
endo_vplot_df = data.frame(Gsdmd = duan.endo[["RNA"]]@data["Gsdmd",],
                           Casp4 = duan.endo[["RNA"]]@data["Casp4",],
                           Irf2 = duan.endo[["RNA"]]@data["Irf2",],
                           Ninj1 = duan.endo[["RNA"]]@data["Ninj1",],
                           Nlrp3 = duan.endo[["RNA"]]@data["Nlrp3",],
                           Casp1 = duan.endo[["RNA"]]@data["Casp1",],
                           Pycard = duan.endo[["RNA"]]@data["Pycard",],
                           Nlrc4 = duan.endo[["RNA"]]@data["Nlrc4",],
                           Aim2 = duan.endo[["RNA"]]@data["Aim2",],
                           Mefv = duan.endo[["RNA"]]@data["Mefv",],
                           Nlrp1b = 0,
                           cluster = duan.endo[["treat"]])
Idents(duan.endo) <- "treat"
log_base = 2
```

```{r}
out_dir <- file.path('../data/Duan_2018/figs/goi_violin')
if(!dir.exists(out_dir)){dir.create(out_dir)}
```
```{r}
feature = 'Gsdmd'
logfc <- (avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "LPS")) - 
  avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "Saline")))/log(log_base)

p <- ggviolin(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  stat_compare_means(method = "wilcox", method.args = list(alternative = "two.sided"), comparisons = list(c("LPS","Saline")), 
                     label = "p.signif", label.y = 2.3, bracket.size = 1, size = 5) +
  ggtitle(label = feature) +
  ylim(c(0,2.4)) +
  annotate("text", x = 2.3, y = 2.4, label=bquote(paste('Log'[2]*'FC = ', .(round(logfc, digits = 3)))), size = 5) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_text(size = 16),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_text(size = 20),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.tiff"), plot = p, device = "tiff", path = out_dir, width = 6, height = 4)

ggviolin(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  ylim(c(0,2.4)) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_blank(),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_blank(),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.clean.tiff"), device = "tiff", path = out_dir, width = 5.55, height = 5)
```
```{r}
feature = "Casp4"
logfc <- (avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "LPS")) - 
  avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "Saline")))/log(log_base)

?desc_statby(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  stat_compare_means(method = "wilcox", method.args = list(alternative = "two.sided"), comparisons = list(c("LPS","Saline")), 
                     label = "p.signif", label.y = 3.1, bracket.size = 1, size = 5) +
  ggtitle(label = feature) +
  ylim(c(0,3.2)) +
  annotate("text", x = 2.3, y = 3.2, label=bquote(paste('Log'[2]*'FC = ', .(round(logfc, digits = 3)))), size = 5) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_text(size = 16),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_text(size = 20),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.tiff"), device = "tiff", path = out_dir, width = 6, height = 4)

ggviolin(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  ylim(c(0,3.2)) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_blank(),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_blank(),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.clean.tiff"), device = "tiff", path = out_dir, width = 5.55, height = 5)
```
```{r}
feature = "Irf2"
logfc <- (avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "LPS")) - 
  avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "Saline")))/log(log_base)

p <- ggviolin(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  stat_compare_means(method = "wilcox", method.args = list(alternative = "two.sided"), comparisons = list(c("LPS","Saline")), 
                     label = "p.signif", label.y = 3.1, bracket.size = 1, size = 5) +
  ggtitle(label = feature) +
  ylim(c(0,3.2)) +
  annotate("text", x = 2.3, y = 3.2, label=bquote(paste('Log'[2]*'FC = ', .(round(logfc, digits = 3)))), size = 5) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_text(size = 16),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_text(size = 20),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.jpeg"), plot = p, device = "jpeg", path = out_dir, width = 6, height = 5)
```
```{r}
feature = "Ninj1"
logfc <- (avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "LPS")) - 
  avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "Saline")))/log(log_base)

p <- ggviolin(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  stat_compare_means(method = "wilcox", method.args = list(alternative = "two.sided"), comparisons = list(c("LPS","Saline")), 
                     label = "p.signif", label.y = 3.1, bracket.size = 1, size = 5) +
  ggtitle(label = feature) +
  ylim(c(0,3.2)) +
  annotate("text", x = 2.3, y = 3.2, label=bquote(paste('Log'[2]*'FC = ', .(round(logfc, digits = 3)))), size = 5) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_text(size = 16),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_text(size = 20),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.jpeg"), plot = p, device = "jpeg", path = out_dir, width = 6, height = 4)
```
```{r}
feature = 'Nlrp3'
logfc <- (avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "LPS")) - 
  avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "Saline")))/log(log_base)

p <- ggviolin(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  stat_compare_means(method = "wilcox", method.args = list(alternative = "two.sided"), comparisons = list(c("LPS","Saline")), 
                     label = "p.signif", label.y = 2.3, bracket.size = 1, size = 5) +
  ggtitle(label = feature) +
  ylim(c(0,2.4)) +
  annotate("text", x = 2.3, y = 2.4, label=bquote(paste('Log'[2]*'FC = ', .(round(logfc, digits = 3)))), size = 5) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_text(size = 16),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_text(size = 20),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.tiff"), plot = p, device = "tiff", path = out_dir, width = 6, height = 4)

ggviolin(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  ylim(c(0,2.4)) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_blank(),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_blank(),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.clean.tiff"), device = "tiff", path = out_dir, width = 5.55, height = 5)
```
```{r}
feature = 'Casp1'
logfc <- (avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "LPS")) - 
  avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "Saline")))/log(log_base)

p <- ggviolin(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  stat_compare_means(method = "wilcox", method.args = list(alternative = "two.sided"), comparisons = list(c("LPS","Saline")), 
                     label = "p.signif", label.y = 2.3, bracket.size = 1, size = 5) +
  ggtitle(label = feature) +
  ylim(c(0,2.4)) +
  annotate("text", x = 2.3, y = 2.4, label=bquote(paste('Log'[2]*'FC = ', .(round(logfc, digits = 3)))), size = 5) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_text(size = 16),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_text(size = 20),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.tiff"), plot = p, device = "tiff", path = out_dir, width = 6, height = 4)

ggviolin(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  ylim(c(0,2.4)) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_blank(),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_blank(),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.clean.tiff"), device = "tiff", path = out_dir, width = 5.55, height = 5)
```
```{r}
feature = 'Pycard'
logfc <- (avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "LPS")) - 
  avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "Saline")))/log(log_base)

p <- ggviolin(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  stat_compare_means(method = "wilcox", method.args = list(alternative = "two.sided"), comparisons = list(c("LPS","Saline")), 
                     label = "p.signif", label.y = 2.3, bracket.size = 1, size = 5) +
  ggtitle(label = feature) +
  ylim(c(0,2.4)) +
  annotate("text", x = 2.3, y = 2.4, label=bquote(paste('Log'[2]*'FC = ', .(round(logfc, digits = 3)))), size = 5) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_text(size = 16),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_text(size = 20),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.tiff"), plot = p, device = "tiff", path = out_dir, width = 6, height = 4)

ggviolin(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  ylim(c(0,2.4)) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_blank(),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_blank(),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.clean.tiff"), device = "tiff", path = out_dir, width = 5.55, height = 5)
```
```{r}
feature = 'Nlrc4'
logfc <- (avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "LPS")) - 
  avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "Saline")))/log(log_base)

p <- ggviolin(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  stat_compare_means(method = "wilcox", method.args = list(alternative = "two.sided"), comparisons = list(c("LPS","Saline")),
                     label = "p.signif", label.y = 2.3, bracket.size = 1, size = 5) +
  ggtitle(label = feature) +
  ylim(c(0,2.4)) +
  annotate("text", x = 2.3, y = 2.4, label=bquote(paste('Log'[2]*'FC = ', .(round(logfc, digits = 3)))), size = 5) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_text(size = 16),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_text(size = 20),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.tiff"), plot = p, device = "tiff", path = out_dir, width = 6, height = 4)

ggviolin(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  ylim(c(0,2.4)) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_blank(),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_blank(),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.clean.tiff"), device = "tiff", path = out_dir, width = 5.55, height = 5)
```
```{r}
feature = 'Aim2'
logfc <- (avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "LPS")) - 
  avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "Saline")))/log(log_base)

p <- ggviolin(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  stat_compare_means(method = "wilcox", method.args = list(alternative = "two.sided"), comparisons = list(c("LPS","Saline")),
                     label = "p.signif", label.y = 2.3, bracket.size = 1, size = 5) +
  ggtitle(label = feature) +
  ylim(c(0,2.4)) +
  annotate("text", x = 2.3, y = 2.4, label=bquote(paste('Log'[2]*'FC = ', .(round(logfc, digits = 3)))), size = 5) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_text(size = 16),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_text(size = 20),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.tiff"), plot = p, device = "tiff", path = out_dir, width = 6, height = 4)

ggviolin(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  ylim(c(0,2.4)) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_blank(),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_blank(),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.clean.tiff"), device = "tiff", path = out_dir, width = 5.55, height = 5)
```
```{r}
feature = 'Mefv'
logfc <- (avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "LPS")) - 
  avgLogExp(duan.endo, feature = feature, cells = WhichCells(duan.endo, idents = "Saline")))/log(log_base)

p <- ggviolin(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  stat_compare_means(method = "wilcox", method.args = list(alternative = "two.sided"), comparisons = list(c("LPS","Saline")),
                     label = "p.signif", label.y = 2.3, bracket.size = 1, size = 5) +
  ggtitle(label = feature) +
  ylim(c(0,2.4)) +
  annotate("text", x = 2.3, y = 2.4, label=bquote(paste('Log'[2]*'FC = ', .(round(logfc, digits = 3)))), size = 5) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_text(size = 16),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_text(size = 20),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.tiff"), plot = p, device = "tiff", path = out_dir, width = 6, height = 4)

ggviolin(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("jitter","boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  ylim(c(0,2.4)) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_blank(),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_blank(),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.clean.tiff"), device = "tiff", path = out_dir, width = 5.55, height = 5)
```
```{r}
feature = 'Nlrp1b'
logfc <- "-"

p <- ggboxplot(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  stat_compare_means(method = "wilcox", method.args = list(alternative = "two.sided"), comparisons = list(c("LPS","Saline")),
                     label = "p.signif", label.y = 2.3, bracket.size = 1, size = 5) +
  ggtitle(label = feature) +
  ylim(c(0,2.4)) +
  annotate("text", x = 2.3, y = 2.4, label=bquote(paste('Log'[2]*'FC = -'))) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_text(size = 16),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_text(size = 20),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.tiff"), plot = p, device = "tiff", path = out_dir, width = 6, height = 4)

ggboxplot(endo_vplot_df, x = "treat", y = feature, fill = "treat", palette = cols[c(7,10)],
         add = c("boxplot"), add.params = list(fill = "white", width = 0.1, alpha = 0.3),
         width = 0.6, size = 1.2) +
  ylim(c(0,2.4)) +
  theme(legend.position = "none",
        axis.line.x = element_line(color = "gray"),
        axis.text.x = element_blank(),
        axis.ticks = element_line(color = "gray"),
        axis.title = element_blank(),
        axis.line.y = element_line(size = 1.2, color = "gray"),
        axis.text.y = element_blank(),
        plot.title = element_text(hjust = 0.5, face = "bold.italic", size = 18))
ggsave(filename = paste0("violinplot_",feature,"_psig_lfc.clean.tiff"), device = "tiff", path = out_dir, width = 5.55, height = 5)
```

#### Augur

```{r}
duan.augur.df <- readRDS("../data/Duan_2018/duan.augur.df.rds")
```

Iteratively compute differentially expressed genes.
```{r}
degs <- data.frame()
for(cells in levels(duan.seurat$cell_type_brief)){
  degs <- FindMarkers(duan.seurat, 
              ident.1 = WhichCells(duan.seurat, expression = cell_type_brief == cells & treat == "LPS"),
              ident.2 = WhichCells(duan.seurat, expression = cell_type_brief == cells & treat == "Saline")
              ) %>% 
    dplyr::filter(p_val_adj <= 0.01) %>% ## Default logfc (base=exp(1)) threshold 0.25
    dplyr::mutate(cell_type = cells) %>% tibble::rownames_to_column(var = "gene") %>%
    dplyr::bind_rows(degs)
}
```

```{r}
duan.augur.df <- dplyr::count(degs, cell_type) %>% 
  dplyr::rename(n_degs = n) %>% 
  dplyr::right_join(duan.augur.df, by = "cell_type")
duan.augur.df$cell_type <- as.factor(duan.augur.df$cell_type)
```

Parse color palette used.
```{r}
DimPlot(duan.seurat, group.by = "cell_type_brief", reduction = "tsne.int", label = TRUE) + 
  NoAxes()
```
```{r}
## Create vector of default ggplot2 colors
## https://github.com/satijalab/seurat/issues/257
my_color_palette <- scales::hue_pal()(length(unique(duan.seurat$cell_type_brief)))
scales::show_col(my_color_palette)

## https://stackoverflow.com/questions/51867716/plot-colors-with-hex-values-in-r/51867915
col_df <- data.frame(
  hex = my_color_palette,
  ct = levels(duan.seurat$cell_type_brief),
  col = c(1,2,3,4,1,2,3,4,1,2,3,4,1,2,3),
  row = c(1,1,1,1,2,2,2,2,3,3,3,3,4,4,4)
)
col_df$hex_ct <- paste(col_df$hex, col_df$ct, sep = "\n")
ggplot(col_df, aes(x = col, y = row, fill = hex)) +
  geom_tile() +
  geom_text(aes(label = hex_ct)) +
  scale_fill_identity() +
  theme_minimal()
ggsave("t-SNE_colors.jpeg", device = "jpeg", path = "../data/Duan_2018/figs/", width = 8, height = 8)
```
```{r}
col_pal.df <- as.data.frame(
  cbind(
  my_color_palette,
  c("SMC","FB","Peri","RBCs","OPCs","Astro","Neurons","Endo","OL","CPEpiC","CR","MG","Interneurons","Epen","MΦ")
))
colnames(col_pal.df) <- c("col_pal","cell_type")
duan.augur.df <- dplyr::mutate(duan.augur.df, cell_type = fct_reorder(cell_type, auc))
```
```{r}
duan.augur.df <- dplyr::right_join(duan.augur.df, col_pal.df, by = "cell_type")
duan.augur.df <- dplyr::mutate(duan.augur.df, cell_type = fct_reorder(cell_type, auc))
```

```{r}
saveRDS(duan.augur.df, file = "../data/Duan_2018/duan.augur.df.rds")
```

```{r}
ggplot(duan.augur.df, mapping = aes(x = cell_type, y = auc)) +
  coord_flip() +
  geom_point(aes(fill = cell_type), size=3, shape=21, stroke=0) +
  geom_segment(aes(x=cell_type, xend=cell_type, y=0, yend=auc, color = cell_type), size = 1) +
  scale_fill_manual(values = duan.augur.df$col_pal,
                    breaks = duan.augur.df$cell_type) +
  scale_color_manual(values = duan.augur.df$col_pal,
                     breaks = duan.augur.df$cell_type) +
  geom_text(mapping = aes(label = round(auc, digits = 2)), hjust = -0.32, size = 3) +
  ylim(c(0.0, 1.2)) +
  scale_y_continuous(
    limits = c(0, 1),
    expand = c(0, 0)) +  # Force Y-axis start from 0
  # scale_y_continuous(labels = c(0.0,0.2,0.4,0.8,1.0)) + 
  xlab(label = "Cell type") +
  ylab(label = "Augur AUC") +
  theme_bw() +
  theme(
    panel.grid.major.x = element_blank(),
    panel.border = element_rect(size = 1.6, fill = NA),
    axis.ticks.x = element_blank(),
    axis.text = element_text(size = 10),
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.title = element_text(size = 12),
    legend.position = "none"
    )
```
```{r}
ggsave(filename = "duan.lollipop.augur.tiff", device = "tiff", path = "../data/Duan_2018", width = 80, height = 100, dpi = 300, units = "mm", family = "Arial")
```

Clean plot with no text on it.
```{r}
ggplot(duan.augur.df, mapping = aes(x = cell_type, y = auc)) +
  coord_flip() +
  geom_point(aes(fill=cell_type), size=3, shape=21, stroke=0) +
  geom_segment(aes(x=cell_type, xend=cell_type, y=0, yend=auc, color = cell_type), size = 1) +
  scale_fill_manual(values = duan.augur.df$col_pal,
                    breaks = duan.augur.df$cell_type) +
  scale_color_manual(values = duan.augur.df$col_pal,
                     breaks = duan.augur.df$cell_type) +
  # geom_text(mapping = aes(label = round(auc, digits = 2)), hjust = -0.32, size = 3) +
  ylim(c(0.0, 1.2)) +
  scale_y_continuous(
    limits = c(0, 1),
    expand = c(0, 0)) +  # Force Y-axis start from 0
  # scale_y_continuous(labels = c(0.0,0.2,0.4,0.8,1.0)) + 
  xlab(label = NULL) +
  ylab(label = NULL) +
  theme_bw() +
  theme(
    panel.grid.major.x = element_blank(),
    panel.border = element_rect(size = 1.6, fill = NA),
    axis.ticks.x = element_blank(),
    axis.text = element_blank(),
    # axis.text.x = element_text(angle = 45, hjust = 1),
    axis.title = element_text(size = 12),
    legend.position = "none"
    )
```
```{r}
ggsave(filename = "duan.lollipop.augur.clean.tiff", device = "tiff", path = "../data/Duan_2018", width = 80, height = 100, dpi = 300, units = "mm")
```

```{r}
pearson.res <- cor.test(duan.augur.df$n_degs, duan.augur.df$auc, method = "pearson", conf.level = 0.95)
ggplot(duan.augur.df, aes(x = n_degs, y = auc)) +
  geom_point(aes(color = cell_type), size=3) +
  scale_color_manual(values = duan.augur.df$col_pal,
                    breaks = duan.augur.df$cell_type) +
  geom_smooth(method="lm", se = FALSE, linetype = "dashed") +
  # geom_text_repel(aes(label = cell_type), box.padding = 1, size = 4, force = 5) +
  geom_label_repel(data = subset(duan.augur.df, cell_type %in% c("Endo")), aes(label = cell_type), 
                   nudge_x = -30, nudge_y = 0.04, box.padding = 0.25, point.padding = 0.4, size = 4, force = 5) +
  geom_label_repel(data = subset(duan.augur.df, cell_type %in% c("MΦ")), aes(label = "MΦ"),
                   nudge_y = 0.04, box.padding = 0.25, point.padding = 0.4, size = 4, force = 5) +
  geom_label_repel(data = subset(duan.augur.df, cell_type %in% c("Peri")), aes(label = cell_type),
                   nudge_x = -20, nudge_y = 0.05, box.padding = 0.25, point.padding = 0.4, size = 4, force = 5) +
  # geom_label_repel(data = subset(duan.augur.df, cell_type %in% c("Cort")), aes(label = cell_type), 
  #                  nudge_x = -30, nudge_y = -0.05, box.padding = 0.25, point.padding = 0.4, size = 4, force = 5) +
  # geom_label_repel(data = subset(duan.augur.df, cell_type %in% c("Lac")), aes(label = cell_type),
  #                  nudge_x = 50,nudge_y = -0.04, box.padding = 0.25, point.padding = 0.4, size = 4, force = 5) +
  xlab("Number of DE genes") +
  ylab("Augur AUC") +
  scale_x_continuous(minor_breaks = seq(0,500,100)) +
  scale_y_continuous(minor_breaks = seq(0,0.9,0.05)) +
  annotate("text", x = 180, y = 0.89, 
           label = paste0("R = ",round(pearson.res$estimate, digits = 3),", p = ", sprintf("%.5f",pearson.res$p.value)), 
           color = "blue") +
  theme(
    panel.border = element_rect(size = 1.6, fill = NA),
    panel.background = element_blank(),
    panel.grid = element_line(color = "#CCCCCC", size = .6),
    axis.text = element_text(size = 10),
    axis.title = element_text(size = 13),
    legend.position = "none"
  )
```
```{r}
ggsave(filename = "duan.scatter.augur.tiff", device = "tiff", path = "../data/Duan_2018", width = 150, height = 120, dpi = 300, units = "mm", family = "Arial")
```

Clean plot with no text on it.
```{r}
ggplot(duan.augur.df, aes(x = n_degs, y = auc)) +
  geom_point(aes(color = cell_type), size=3) +
  scale_color_manual(values = duan.augur.df$col_pal,
                    breaks = duan.augur.df$cell_type) +
  # geom_smooth(method="lm", se = FALSE, linetype = "dashed") +
  # geom_text_repel(aes(label = cell_type), box.padding = 1, size = 4, force = 5) +
  # geom_label_repel(data = subset(duan.augur.df, cell_type %in% c("Endo")), aes(label = cell_type), 
  #                  nudge_x = -30, nudge_y = 0.04, box.padding = 0.25, point.padding = 0.4, size = 4, force = 5) +
  # geom_label_repel(data = subset(duan.augur.df, cell_type %in% c("MΦ")), aes(label = "MΦ"),
  #                  nudge_y = 0.04, box.padding = 0.25, point.padding = 0.4, size = 4, force = 5) +
  # geom_label_repel(data = subset(duan.augur.df, cell_type %in% c("Peri")), aes(label = cell_type),
  #                  nudge_x = -20, nudge_y = 0.05, box.padding = 0.25, point.padding = 0.4, size = 4, force = 5) +
  # geom_label_repel(data = subset(duan.augur.df, cell_type %in% c("Cort")), aes(label = cell_type), 
  #                  nudge_x = -30, nudge_y = -0.05, box.padding = 0.25, point.padding = 0.4, size = 4, force = 5) +
  # geom_label_repel(data = subset(duan.augur.df, cell_type %in% c("Lac")), aes(label = cell_type),
  #                  nudge_x = 50,nudge_y = -0.04, box.padding = 0.25, point.padding = 0.4, size = 4, force = 5) +
  xlab(label = NULL) +
  ylab(label = NULL) +
  # scale_x_continuous(minor_breaks = seq(0,500,100)) +
  scale_y_continuous(limits = c(0.43,0.9), minor_breaks = seq(0,0.9,0.05)) +
  # annotate("text", x = 180, y = 0.89, 
  #          label = paste0("R = ",round(pearson.res$estimate, digits = 3),", p = ", sprintf("%.5f",pearson.res$p.value)), 
  #          color = "blue") +
  theme(
    panel.border = element_rect(size = 1.6, fill = NA),
    panel.background = element_blank(),
    panel.grid = element_line(color = "#CCCCCC", size = .6),
    axis.text = element_blank(),
    axis.title = element_text(size = 13),
    legend.position = "none"
  )
```
```{r}
ggsave(filename = "duan.scatter.augur.clean.tiff", device = "tiff", path = "../data/Duan_2018", width = 150, height = 120, dpi = 300, units = "mm")
```

## GO analysis

```{r}
degs$entrez <- mapIds(org.Mm.eg.db, keys = degs$gene, column = "ENTREZID", keytype = "SYMBOL", multiVals = "first")
input_genes <- degs %>% dplyr::filter(cell_type == "Endo") %>% dplyr::arrange(desc(avg_log2FC)) %>% dplyr::slice(1:205) %>% dplyr::pull(entrez)
background <- mapIds(org.Mm.eg.db, keys = rownames(duan.endo), column = "ENTREZID", keytype = "SYMBOL", multiVals = "first")
```

```{r}
de_ego.BP <- enrichGO(gene     = input_genes,
                      universe = background,
                      OrgDb    = org.Mm.eg.db,
                      ont      = "BP",
                      pAdjustMethod = "BH",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05,
                      readable = TRUE)
View(de_ego.BP@result)
write.csv(de_ego.BP@result, file = "../data/Duan_2018/Endo_GO_BP.csv", quote = FALSE)
```

```{r}
# GO_BP_terms_drop <- c("GO:0051607","GO:0031347","GO:0044403","GO:0002237","GO:1903901","GO:0032496","GO:0002831","GO:0048525","GO:0071222",
#                       "GO:0071219","GO:0045071","GO:0002250","GO:1903900","GO:0071216","GO:0002460","GO:0043903")
# de_ego.BP.dropped <- dropGO(de_ego.BP, term = GO_BP_terms_drop)
GO_BP_terms_use <- c("GO:0009617","GO:0009615","GO:0034341","GO:0002697","GO:0035456","GO:0002250","GO:0019221","GO:0050727")
```
```{r}
GO_BP_plot <- de_ego.BP@result %>% dplyr::filter(ID %in% GO_BP_terms_use) %>% dplyr::arrange(p.adjust) %>% 
  mutate(Description = fct_reorder(Description, dplyr::desc(p.adjust)))
ggplot(GO_BP_plot, aes(x = Description, y = -log10(p.adjust), fill = Count)) +
  coord_flip() +
  ylab(bquote(-log[10](p.adj))) +
  geom_bar(stat = "identity", color = "#333333", width = 0.67) +
  scale_y_continuous(expand = c(0,0),
                     limits = c(0,36)) +
  scale_fill_continuous(low = "brown", high = "orange") +
  theme_bw() +
  theme(axis.title.y = element_blank(),
        axis.text.y = element_text(size = 12),
        axis.title.x = element_text(size = 14),
        axis.text.x = element_text(size = 10),
        legend.text = element_text(size = )
        )
```
```{r}
ggsave(filename = "GO_BP_endo.tiff", device = "tiff", path = "../data/Duan 2018"), width = 180, height = 96, dpi = 300, units = "mm")
```

```{r}
ggplot(GO_BP_plot, aes(x = Description, y = -log10(p.adjust), fill = Count)) +
  coord_flip() +
  ylab(bquote(-log[10](p.adj))) +
  geom_bar(stat = "identity", color = "#333333", width = 0.67) +
  scale_y_continuous(expand = c(0,0),
                     limits = c(0,36)) +
  scale_fill_continuous(low = "brown", high = "orange") +
  theme_bw() +
  theme(axis.title.y = element_blank(),
        axis.text.y = element_blank(),
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        legend.text = element_blank(),
        legend.title = element_blank()
        )
```
```{r}
ggsave(filename = "GO_BP_endo.clean.tiff", device = "tiff", path = file.path("../data/Duan 2018"), width = 180, height = 96, dpi = 300, units = "mm")
```

## Heatmap plot of Endo DEGs

```{r}
gois_200 <- degs %>% dplyr::filter(cell_type == "Endo") %>% dplyr::arrange(desc(avg_log2FC)) %>% 
                                      dplyr::slice(1:205) %>% dplyr::pull(gene)
duan.endo <- ScaleData(duan.endo, 
                       features = c(VariableFeatures(duan.endo), gois_200))
```
```{r}
# h <- DoHeatmap(duan.endo, 
#           features = gois_200, group.by = "treat", label = FALSE) +
#   scale_fill_distiller(palette = 'RdGy') +
#   NoAxes()
h_clean <- DoHeatmap(duan.endo, 
          features = gois_200, group.by = "treat", label = FALSE, group.colors = cols[c(7,10)]) +
  scale_fill_distiller(palette = 'RdGy') +
  NoAxes() +
  theme(legend.text = element_blank(),
        legend.title = element_blank())
```
```{r}

ggsave(filename = "duan.heatmap.clean.eps", device = "eps", plot = h_clean, path = "../data/Duan_2018", width = 150, height = 120, dpi = 300, units = "mm")
ggsave(filename = "duan.heatmap.clean.tiff", device = "tiff", plot = h_clean, path = "../data/Duan_2018", width = 150, height = 120, dpi = 300, units = "mm")
```

#### Fig Supplementary Canonical Marker Expression

```{r}
canonical_markers <- list(
  Neurons = c("Rbfox2"),
  Interneurons = c("Gad2"),
  `Cajal-Retzius Cells` = c("Ndnf"),
  Astrocytes = c("Aqp4"),
  `Ependymal Cells` = c("Ccdc153"),
  `Endothelial Cells` = c("Cldn5", "Pecam1"),
  Pericytes = c("Kcnj8", "Anpep"),
  SMCs = c("Acta2"),
  Fibroblasts = c("Col1a1", "Lum"),
  Microglia = c("Cx3cr1", "Aif1"),
  Macrophage = c("Mrc1"),
  `Red Blood Cells` = c("Hbb-bt"),
  `Oligodendrocyte Precursor Cells` = c("Pdgfra"),
  Oligodendrocytes = c("Mbp"),
  `Choroid Plexus Epithelial Cells` = c("Ttr")
  )

for (ct in names(canonical_markers)){
  for (mk in canonical_markers[[ct]]){
    FeaturePlot(duan.seurat, features = mk, reduction = "tsne.int") + 
      scale_colour_gradientn(colours = brewer.pal(n = 9, name = "RdPu")) +
      xlab("TSNE_1") + 
      ylab("TSNE_2") + 
      theme(
        plot.title = element_text(hjust = 0.02, vjust = -8),
        panel.border = element_rect(color = "black"),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        legend.position = c(0.9,0.18)
        )
    ggsave(filename = paste0(ct,"_",mk,".tiff"), device = "tiff", 
           path = "../data/Duan_2018/figs/canonical_markers/", 
           dpi = 300, width = 6, height = 5)
    
    FeaturePlot(duan.seurat, features = mk, reduction = "tsne.int") + 
      scale_colour_gradientn(colours = brewer.pal(n = 9, name = "RdPu")) +
      ggtitle(NULL) +
      xlab(NULL) + 
      ylab(NULL) + 
      theme(
        panel.border = element_rect(color = "black"),
        axis.text = element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks = element_blank(),
        legend.position = c(0.9,0.16),
        legend.key.size = unit(12, units = "pt"),
        legend.text = element_blank()
        )
    ggsave(filename = paste0(ct,"_",mk,"_clean.tiff"), device = "tiff", 
           path = "../data/Duan_2018/figs/canonical_markers/", 
           dpi = 300, width = 4, height = 3.6)
  }
}
```
